lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,119 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Dict, Iterator, List, MutableMapping, Union
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, str]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self):
+        # avoid using ._query_params = {} as that will use __setattr__
+        # and __setattr__ accesses ._query_params and ._query_params is not defined yet.
+        # https://stackoverflow.com/a/25179980
+        self.__dict__["_query_params"] = {}
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._query_params.keys())
+
+    def __getitem__(self, key: str) -> str:
+        try:
+            # avoid using ._query_params[key] as that will use __getattr__
+            # which relies on __getitem__ and that causes an infinite loop
+            value = self.__dict__["_query_params"][key]
+            if isinstance(value, list):
+                if len(value) == 0:
+                    return ""
+                else:
+                    return value[-1]  # type: ignore
+            return value  # type: ignore
+        except KeyError:
+            raise KeyError(_missing_key_error_message(key))
+
+    # Type checking users should handle the string serialization themselves
+    # We will accept any type for the list and serialize to str just in case
+    def __setitem__(self, key: str, value: Union[str, List[str]]) -> None:
+        if isinstance(value, list):
+            self._query_params[key] = [str(item) for item in value]
+        else:
+            self._query_params[key] = str(value)
+        self._send_query_param_msg()
+
+    def __delitem__(self, key: str) -> None:
+        if key in self._query_params:
+            del self._query_params[key]
+            self._send_query_param_msg()
+        else:
+            raise KeyError(_missing_key_error_message(key))
+
+    def get_all(self, key: str) -> List[str]:
+        if key not in self._query_params:
+            return []
+        value = self._query_params[key]
+        return value if isinstance(value, list) else [value]
+
+    def __getattr__(self, key: str) -> str:
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(_missing_key_error_message(key))
+
+    def __setattr__(self, key: str, value: str) -> None:
+        try:
+            self[key] = value
+        except KeyError:
+            raise AttributeError(_missing_key_error_message(key))
+
+    def __delattr__(self, key: str) -> None:
+        try:
+            del self[key]
+        except KeyError:
+            raise AttributeError(_missing_key_error_message(key))
```

Comment:
I should have realized this earlier, but do we even want to define these methods here? I'm almost sure the answer is no since we can always make the user-facing API support both key and attr based access in the proxy class (which does require some changes that I previously suggested in a later PR to be reverted, my bad).

Defining these here just makes us run into weird attr-related issues like in the `__init__` method and elsewhere, and it doesn't do anything helpful for us wrt getting the user-facing API be what we want it to be since we can just do that in the proxy class.